### PR TITLE
Medtronic driver merge

### PR DESCRIPTION
**Pull request by andyrozman**
This is implementation of Medtronic driver. For now its hidden under engineering mode. Documentation will be merged soon.

### CODE REVIEW

**Comment by jotomo** [comment]
Some notes from a very quick first look:
TreatmentsPlugin: why all the style changes? Makes it hard to see what has actually changed.
README: still caries the MD-branch specific instructions
FabricPrivacy: still has debug code
logback: revert debug changes?

File: app/src/main/java/info/nightscout/androidaps/plugins/general/overview/dialogs/MessageDialog.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
+package info.nightscout.androidaps.plugins.general.overview.dialogs;
```
Comment (MilosKozak) [inline app/src/main/java/info/nightscout/androidaps/plugins/general/overview/dialogs/MessageDialog.java]: Isn't  it duplicate to `OKDialog`?

File: app/src/main/java/info/nightscout/androidaps/MainApp.java [hunk 2]
```diff
@@ -252,6 +264,54 @@ private void registerLocalBroadcastReceiver() {
         lbm.registerReceiver(dbAccessReciever, new IntentFilter(Intents.ACTION_DATABASE));
     }
 
+
+    private void setBTReceiver() {
```
Comment (MilosKozak) [inline app/src/main/java/info/nightscout/androidaps/MainApp.java]: Can this be standart receiver defined in manifest?

File: app/src/main/java/info/nightscout/androidaps/MainApp.java [hunk 3]
```diff
@@ -252,6 +264,54 @@ private void registerLocalBroadcastReceiver() {
...
+                            case BluetoothAdapter.STATE_ON: {
+                                if ("Medtronic".equals(activePump.deviceID())) {
+                                    Log.v("MainApp", "Bluetooth on");
+                                    RileyLinkUtil.sendBroadcastMessage(RileyLinkConst.Intents.BluetoothReconnected);
+                                }
+                            }
+                            break;
+                        }
+                    } else {
+                        activePump.timeDateOrTimeZoneChanged();
```
Comment (MilosKozak) [inline app/src/main/java/info/nightscout/androidaps/MainApp.java]: How is this related to timezone change?

File: app/src/main/java/info/nightscout/androidaps/plugins/pump/common/defs/PumpType.java.orig [hunk 4]
```diff
@@ -0,0 +1,398 @@
+package info.nightscout.androidaps.plugins.pump.common.defs;
```
Comment (MilosKozak) [inline app/src/main/java/info/nightscout/androidaps/plugins/pump/common/defs/PumpType.java.orig]: redundant?

File: app/src/test/java/info/nightscout/androidaps/utils/LoggerRule.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
+package info.nightscout.androidaps.utils;
```
Comment (MilosKozak) [inline app/src/test/java/info/nightscout/androidaps/utils/LoggerRule.java]: ???

File: app/src/test/java/android/util/Log.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
+package android.util;
```
Comment (MilosKozak) [inline app/src/test/java/android/util/Log.java]: Can this be transformed to use Logger as everywhere else?

File: app/src/main/res/xml/pref_medtronic.xml [hunk 7]
```diff
@@ -0,0 +1,81 @@
+<?xml version="1.0" encoding="utf-8"?>
+<PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android">
+    <PreferenceCategory android:title="@string/medtronic_name">
+
+
+        <EditTextPreference
+            android:defaultValue="000000"
+            android:key="pref_medtronic_serial"
```
Comment (MilosKozak) [inline app/src/main/res/xml/pref_medtronic.xml]: use  `key_...` like everywhere alse to avoid typos

File: app/src/main/res/layout/overview_wizard_dialog_compact.xml [hunk 8]
```diff
@@ -0,0 +1,522 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (MilosKozak) [inline app/src/main/res/layout/overview_wizard_dialog_compact.xml]: not referenced

File: app/src/main/java/info/nightscout/androidaps/plugins/treatments/TreatmentsPlugin.java [hunk 9]
```diff
@@ -550,6 +579,10 @@ public boolean addToHistoryTreatment(DetailedBolusInfo detailedBolusInfo, boolea
         return newRecordCreated;
     }
 
+    private boolean isMedtronicPump() {
```
Comment (MilosKozak) [inline app/src/main/java/info/nightscout/androidaps/plugins/treatments/TreatmentsPlugin.java]: better use `MedtronicPlugin.getInstance().isEnabled(PluginType.PUMP)`

File: app/src/main/java/info/nightscout/androidaps/plugins/pump/virtual/VirtualPumpStatus.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
+package info.nightscout.androidaps.plugins.pump.virtual;
```
Comment (MilosKozak) [inline app/src/main/java/info/nightscout/androidaps/plugins/pump/virtual/VirtualPumpStatus.java]: not used

File: app/src/main/java/info/nightscout/androidaps/MainApp.java (hunk 2 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/MainApp.java]: Maybe. I have no deep understanding how Broadcast receivers work...

File: app/src/main/java/info/nightscout/androidaps/MainApp.java (hunk 3 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/MainApp.java]: I register 4 intents, one is for BT other three are for TZ/Time/Date changes... I wanted to do writing to Db too, but for now this was enough...

File: app/src/main/java/info/nightscout/androidaps/plugins/pump/common/defs/PumpType.java.orig (hunk 4 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/plugins/pump/common/defs/PumpType.java.orig]: what is redundant?

Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/plugins/pump/common/defs/PumpType.java.orig]: I see what you mean. Will fix.

File: app/src/test/java/android/util/Log.java (hunk 6 above)
Comment (andyrozman) [reply app/src/test/java/android/util/Log.java]: This was for Unit Tests.. I am not able to see any Logger results if I run tests... If I override Log java in test, you can at least follow this log entries...

File: app/src/main/java/info/nightscout/androidaps/plugins/pump/virtual/VirtualPumpStatus.java (hunk 10 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/plugins/pump/virtual/VirtualPumpStatus.java]: Will remove...

File: app/src/main/res/layout/overview_wizard_dialog_compact.xml (hunk 8 above)
Comment (andyrozman) [reply app/src/main/res/layout/overview_wizard_dialog_compact.xml]: Will remove

File: app/src/test/java/info/nightscout/androidaps/utils/LoggerRule.java (hunk 5 above)
Comment (andyrozman) [reply app/src/test/java/info/nightscout/androidaps/utils/LoggerRule.java]: Will remove... One of several tries to get Logger to start printing logs when running tests...

File: app/src/main/java/info/nightscout/androidaps/plugins/general/overview/dialogs/MessageDialog.java (hunk 1 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/plugins/general/overview/dialogs/MessageDialog.java]: I will try to use OKDialog, and if it works same, we can remove this...

Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/plugins/general/overview/dialogs/MessageDialog.java]: OKDialog doesn't work... It is just not displayed... I modified MessageHelperActivity to call OKDialog and it just didn't work...

File: app/src/main/java/info/nightscout/androidaps/plugins/treatments/TreatmentsPlugin.java (hunk 9 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/plugins/treatments/TreatmentsPlugin.java]: Would that work always?

File: app/src/main/res/xml/pref_medtronic.xml (hunk 7 above)
Comment (andyrozman) [reply app/src/main/res/xml/pref_medtronic.xml]: Insight doesn't do it...

File: app/src/main/java/info/nightscout/androidaps/plugins/treatments/TreatmentsPlugin.java (hunk 9 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/plugins/treatments/TreatmentsPlugin.java]: it seems it does...

File: app/src/main/java/info/nightscout/androidaps/plugins/pump/common/PumpPluginAbstract.java [hunk 11]
```diff
@@ -0,0 +1,446 @@
...
+import info.nightscout.androidaps.utils.DateUtil;
+import info.nightscout.androidaps.utils.DecimalFormatter;
+
+/**
+ * Created by andy on 23.04.18.
+ */
+
+// When using this class, make sure that your first step is to create mConnection (see MedtronicPumpPlugin)
+
+public abstract class PumpPluginAbstract extends PluginBase implements PumpInterface, ConstraintsInterface {
```
Comment (jotomo) [inline app/src/main/java/info/nightscout/androidaps/plugins/pump/common/PumpPluginAbstract.java]: This class is a MDT-specific base class, it should not be in the _common_ package.

File: app/src/main/java/info/nightscout/androidaps/MainApp.java (hunk 3 above)
Comment (jotomo) [reply app/src/main/java/info/nightscout/androidaps/MainApp.java]: This mixes too many things. Calling it a BTReceiver when it does a whole lot more is confusing :-)
While we don't have clean plugin/pump lifecycles yet (Combo is very guilty of that, binding to ruffy service on app start), we should aim for it and separate pump plugins. 
This receiver will trigger Riley code no matter which pump driver is active (and couple MainApp to a RileyLink impl class). I would suggest moving the BT-receiver to MDT-land and register it dynamically (not in the manifest, since it's not always needed) in _onStart_ and unregister in _onStop_.
Handling timezone / date changes _might_ be something to be done on a level above pump drivers and just have the drivers provide a _setDateTime_ method, but that probably won't work since drivers handle time changes differently (this should be discussed, but not as part of a big merge like this). So keeping a receiver here (registered in manifest) to receive time changes and informing the drivers seems a good place to start.

File: app/src/main/java/info/nightscout/androidaps/utils/SP.java [hunk 12]
```diff
@@ -132,6 +132,19 @@ static public void putInt(int resourceID, int value) {
...
+        editor.putString(MainApp.gs(resourceID), value.toString());
+        editor.apply();
+    }
+
+    static public void putDouble(String key, Double value) {
+        SharedPreferences.Editor editor = sharedPreferences.edit();
+        editor.putString(key, value.toString());
+        editor.apply();
+    }
+
```
Comment (jotomo) [inline app/src/main/java/info/nightscout/androidaps/utils/SP.java]: Can be removed, they're not used (they're overloading version with a primitive double as value) and will fail for null-values.

File: app/src/main/java/info/nightscout/androidaps/plugins/general/overview/dialogs/MessageDialog.java (hunk 1 above)
Comment (jotomo) [reply app/src/main/java/info/nightscout/androidaps/plugins/general/overview/dialogs/MessageDialog.java]: Locking at invocation sites ... is the helper activity needed? Callers already open a new thread, so there's no blocking issue and _OkDialog.show()_ (not sure if this blocks) could be invoked directly.
However, the invocation in _MedtronicUtil.displayNotConfiguredDialog_:  messages like these are usually displayed as notifications on overview (e.g. _ComboPlugin.checkPumpTime_)
The invocation at _MedtronicPumpPlugin.deliverBolus()_: shouldn't this integrate with _BolusProgressDialog_ since it happens during bolusing? (Don't put time into issues you might see with that dialog, esp. with SDK upgrades this dialog has nowhere to go but being replaced with a progress notification).
With that, neither helper nor _MessageDialog_ itself would be needed anymore if I'm not overlooking something.

Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/plugins/general/overview/dialogs/MessageDialog.java]: displayNotConfiguredDialog - This is for dialogs and actions. We have several dialogs and actions, and calling them when device was not configured produced Exceptions... Putting warning into Overview would be useless here, because User is not in Overview, so he wouldn't see warning until next time. I was testing situation with Bolus, where we also call this dialog, and OKDialog doesn't work there... 
deliverBolus - To get full picture you need to look at method deliverTreatment in PumpPluginAbstract... that one actually calls deliverBolus...

File: app/src/main/java/info/nightscout/androidaps/MainApp.java (hunk 3 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/MainApp.java]: I tried to register it in the pump itself and it just didn't work there... None of BT events was received... Maybe I did mistake somewhere or forgot to register something, but all BT events were ignored when run from Driver itself, I tried several other options (I was playing around with this for few days). It started working when I registered it in MainApp, similar as you do all other Receivers. I could have made some mistake somewhere, I am relatively new to Android programming. 

As for time changes... That is intended for *any* pump class. I extended PumpInterface with timeDateOrTimeZoneChanged() method. It has empty implementation for all drivers at the moment, but this is the point where driver developers can do something when time changes.. We could do separate receiver here or something, I am just big fan or reusing stuff, instead of creating new instances of same stuff that waste memory...

File: app/src/main/java/info/nightscout/androidaps/plugins/general/overview/dialogs/MessageDialog.java (hunk 1 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/plugins/general/overview/dialogs/MessageDialog.java]: That method deliverTreatment was taken from Combo implementation and refactored so that deliverBolus is called (in actual pump driver) when real bolus delivering is going on, and other options in other cases (when no real bolus will be triggered).

Comment (jotomo) [reply app/src/main/java/info/nightscout/androidaps/plugins/general/overview/dialogs/MessageDialog.java]: https://github.com/andyrozman/AndroidAPS/pull/5

File: app/src/main/java/info/nightscout/androidaps/MainApp.java (hunk 3 above)
Comment (jotomo) [reply app/src/main/java/info/nightscout/androidaps/MainApp.java]: Re: time changes: definitely reuse code. My thought was whether it might make sense to also pull the adjustment logic from drivers to a level above, stuff like gradually adjusting time or deciding the jump is to big and annoy the user. Extracting that logic to be above the drivers and have the drivers implement dumber logic that only sets the time. I only wish I could set the time on the combo programmatically ;-) In short: that change is fine, I was just derailing the conversation a bit with ponderings going a bit further.

Re: broadcast receivers: not sure what they require to work or setup, still learning Android stuff myself. We have the DataReceiver as a sample for receiving broadcast ... If it needs to stay in MainApp I'd prefer the code to send out an event to which RileyLink etc can subscribe to, rather than have a hard dependency to it in MainApp.

File: app/src/main/java/info/nightscout/androidaps/plugins/pump/common/PumpPluginAbstract.java (hunk 11 above)
Comment (MilosKozak) [reply app/src/main/java/info/nightscout/androidaps/plugins/pump/common/PumpPluginAbstract.java]: 👍

File: app/src/main/java/info/nightscout/androidaps/MainApp.java (hunk 3 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/MainApp.java]: time changes:
Yeah, I was thinking to store this time/date/TZ changes into database... That would make synchronization of data much easier (if pump would need it)... 

Receivers:
I tried using them in Manifest and that didn't work correctly... Maybe I did something wrong I am not sure... So I reverted the change... I could make BroadcastReceiver for both: BT and Date/TZ changes and then just create them in MainApp... That would make it little cleaner...

File: app/src/main/java/info/nightscout/androidaps/plugins/general/overview/dialogs/MessageDialog.java (hunk 1 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/plugins/general/overview/dialogs/MessageDialog.java]: Merged and we can resolve this

File: app/src/main/res/xml/pref_medtronic.xml (hunk 7 above)
Comment (andyrozman) [reply app/src/main/res/xml/pref_medtronic.xml]: So is this required to do? I don't think we gain much to add key into string, but you are the experts... 
I don't use this keys directly in code anywhere (I use MedtronicConst and RileyLinkConst classes)

File: app/src/main/java/info/nightscout/androidaps/MainApp.java (hunk 2 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/MainApp.java]: No. Doesn't work if defined in Manifest.

File: app/src/main/java/info/nightscout/androidaps/plugins/pump/common/PumpPluginAbstract.java (hunk 11 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/plugins/pump/common/PumpPluginAbstract.java]: Its not MDT specific... its not any type of pump specific... its abstract class for pump interface...

File: app/src/main/java/info/nightscout/androidaps/utils/SP.java (hunk 12 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/utils/SP.java]: removed xxxDouble methods not used.

File: app/src/main/java/info/nightscout/androidaps/MainApp.java (hunk 2 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/MainApp.java]: See next comment about refctoring

File: app/src/main/java/info/nightscout/androidaps/MainApp.java (hunk 3 above)
Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/MainApp.java]: So I refactored this a little... I created 2 broadcast receievers... One is registered from MainApp (with MainApp) - Time Change, and other is registered from RileyLinkService (on create) with MainApp.

So I think this is fixed...

File: app/src/main/res/xml/pref_medtronic.xml (hunk 7 above)
Comment (MilosKozak) [reply app/src/main/res/xml/pref_medtronic.xml]: Using keys prevents typo mistakes. It happend many times in AAPS project

File: app/src/main/java/info/nightscout/androidaps/plugins/pump/common/PumpPluginAbstract.java (hunk 11 above)
Comment (MilosKozak) [reply app/src/main/java/info/nightscout/androidaps/plugins/pump/common/PumpPluginAbstract.java]: but it's used only for MDT right now. i'd move it for now

File: app/src/main/java/info/nightscout/androidaps/plugins/pump/common/data/PumpStatus.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
+package info.nightscout.androidaps.plugins.pump.common.data;
```
Comment (MilosKozak) [inline app/src/main/java/info/nightscout/androidaps/plugins/pump/common/data/PumpStatus.java]: PumpStatus is currently MDT specific too

File: app/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -1391,6 +1391,16 @@
     <string name="description_pump_medtronic">Pump integration for Medtronic, requires RileyLink device and specific Pump Model</string>
 
     <!-- MDT Configuration -->
+    <string name="pref_key_medtronic_serial">pref_medtronic_serial</string>
```
Comment (MilosKozak) [inline app/src/main/res/values/strings.xml]: ` translatable="false"` is missing

File: app/src/main/java/info/nightscout/androidaps/plugins/pump/common/PumpPluginAbstract.java (hunk 11 above)
Comment (jotomo) [reply app/src/main/java/info/nightscout/androidaps/plugins/pump/common/PumpPluginAbstract.java]: From what I've seen it extracts things from MDT, including things that don't apply to all pumps. The _ComboPlugin_ doesn't (directly) use a service connection for example. And TBH, the initialization is a tad too complex for my taste, being stretched out over the constructors of the abstract class and the concrete sublass, fields of both classes, the _onStart_, _onCustomStart_ as well as _initPumpStatusData_ methods (while AAPS triggers a pump init by issuing a READ_STATE command). Not trying to bash your code, the Combo code is also not as clear in that regard as I'd like it to be (a spike to evaluate a refactoring towards a cleaner lifecycle with the Combo is on my agenda).
However, I'm getting a bit carried away as far as the MDT-merge itself is concerned (which I don't want to delay over this issue): I think the lifecycle of pump plugins could do with a rework (to actually only init a pump when a plugin is enabled and not init all pump plugins on app start). For such a refactoring, having a base class that simplifies and helps with implementing a clean lifecycle might well be helpful.

Comment (andyrozman) [reply app/src/main/java/info/nightscout/androidaps/plugins/pump/common/PumpPluginAbstract.java]: I intended to use this same abstract class for at least Omnipod and OmnipodDash... but you are right we need to rethink how this things are done... 

I will be working on Omnipod integration with Bart soon (after we finish the merge) and I don't want to duplicate the code where its not needed.

And if everything goes right I will finally get my hands on some Dash pods, so maybe that development will also move from deadpoint...

File: app/src/main/res/values/strings.xml (hunk 14 above)
Comment (andyrozman) [reply app/src/main/res/values/strings.xml]: Will change... give me half an hour...
